vortex-array/src/arrays/listview/rebuild.rs

```diff
@@ -0,0 +1,398 @@
+    ///
+    /// This mode will deduplicate all overlapping list views, such that the [`ListViewArray`] looks
+    /// like a [`ListArray`] but with an additional `sizes` array.
+    ///
+    /// [`ListArray`]: crate::arrays::ListArray
+    MakeZeroCopyToList,
+
+    // TODO(connor)[ListView]: Implement some version of this.
+    /// Finds the shortest packing / overlapping of list elements.
+    ///
+    /// This problem is known to be NP-hard, so maybe when someone proves that P=NP we can implement
+    /// this algorithm (but in all seriousness there are many approximate algorithms that could
+    /// work well here).
+    OverlapCompression,
+}
+
+impl ListViewArray {
+    /// Rebuilds the [`ListViewArray`] according to the specified mode.
+    pub fn rebuild(&self, mode: ListViewRebuildMode) -> ListViewArray {
+        match mode {
+            ListViewRebuildMode::RemoveGaps => self.rebuild_remove_gaps(),
+            ListViewRebuildMode::MakeZeroCopyToList => self.rebuild_zero_copy_to_list(),
+            ListViewRebuildMode::OverlapCompression => unimplemented!("Does P=NP?"),
+        }
+    }
+
+    /// Rebuilds a [`ListViewArray`] by removing unreferenced elements while preserving overlaps.
+    ///
+    /// This removes "garbage" elements that are not referenced by any list. Unlike
+    /// [`rebuild_flatten()`], this preserves overlap structure where multiple lists may reference
+    /// the same elements.
+    ///
+    /// [`rebuild_flatten()`]: Self::rebuild_flatten
+    fn rebuild_remove_gaps(&self) -> ListViewArray {
+        if self.is_empty() {
+            return self.clone();
+        }
+
+        let offset_ptype = self.offsets().dtype().as_ptype();
+        let size_ptype = self.sizes().dtype().as_ptype();
+
+        match_each_integer_ptype!(offset_ptype, |O| {
+            match_each_integer_ptype!(size_ptype, |S| { self.rebuild_remove_gaps_inner::<O, S>() })
+        })
+    }
+
+    fn rebuild_remove_gaps_inner<O, S>(&self) -> ListViewArray
+    where
+        O: IntegerPType,
+        S: IntegerPType,
+    {
+        let offsets_primitive = self.offsets().to_primitive();
+        let sizes_primitive = self.sizes().to_primitive();
+        let offsets_slice = offsets_primitive.as_slice::<O>();
+        let sizes_slice = sizes_primitive.as_slice::<S>();
+
+        // Mark which elements in the elements array are referenced by at least one list.
+        let mut referenced_elements = vec![false; self.elements().len()];
+        for i in 0..self.len() {
+            let offset: usize = offsets_slice[i].as_();
+            let size: usize = sizes_slice[i].as_();
+            for j in offset..offset + size {
+                referenced_elements[j] = true;
+            }
+        }
+
+        // Fast path: if all elements are referenced, no garbage collection needed.
+        if referenced_elements
+            .iter()
+            .all(|&is_referenced| is_referenced)
+        {
+            return self.clone();
+        }
```

Comment:
can you merge this with the previous loop?

Reply:
Yep, you are right

Reply:
Is it worth even checking this? I would assume its non-trival to compute and you are being asked to compact anyway?